src/object.c

```diff
@@ -394,11 +394,14 @@ void dismissSds(sds s) {
     dismissMemory(sdsAllocPtr(s), sdsAllocSize(s));
 }
 
-/* See dismissObject() */
+/* See dismissObject(). String object is an exception. Instead of
+ * dismissing it with madvise(MADV_DONTNEED) we free it via the
+ * allocator. This has advantage that it allows the allocator to
+ * accumulate free buffers to free whole pages. While madvise is nop
+ * if the buffer is less than a page. The allocator overhead is
+ * minimal, because raw strings (sds) use the fast path. */
 void dismissStringObject(robj *o) {
-    if (o->encoding == OBJ_ENCODING_RAW) {
-        dismissSds(o->ptr);
-    }
+    freeStringObject(o);
```

Comment:
maybe there was a good reason, but why don;t we move the logic in dismissSds, this will help capture many other cases of large hashes etc...

Reply:
Tested, `BGREWRITEAOF` behaves the same.